eva/optimizer/rules/rules.py

```diff
@@ -317,6 +319,99 @@ def apply(self, before: LogicalFilter, context: OptimizerContext):
         return new_join_node
 
 
+class XformLateralJoinToLinearFlow(Rule):
+    """If the inner node of a lateral join is a function-valued expression, we
+    eliminate the join node and make the inner node the parent of the outer node. This
+    produces a linear #data flow path. Because this scenario is common in our system,
+    we chose to explicitly convert it to a linear flow, which simplifies the
+    implementation of other optimizations such as UDF reuse and parallelized plans by
+    removing the join."""
+
+    def __init__(self):
+        pattern = Pattern(OperatorType.LOGICALJOIN)
+        pattern.append_child(Pattern(OperatorType.DUMMY))
+        pattern.append_child(Pattern(OperatorType.LOGICALFUNCTIONSCAN))
+        super().__init__(RuleType.XFORM_LATERAL_JOIN_TO_LINEAR_FLOW, pattern)
+
+    def promise(self):
+        return Promise.XFORM_LATERAL_JOIN_TO_LINEAR_FLOW
+
+    def check(self, before: LogicalJoin, context: OptimizerContext):
+        if before.join_type == JoinType.LATERAL_JOIN:
+            if before.join_predicate is None and not before.join_project:
+                return True
+        return False
+
+    def apply(self, before: LogicalJoin, context: OptimizerContext):
+        #     LogicalJoin(Lateral)              LogicalApplyAndMerge
+        #     /           \                 ->       |
+        #    A        LogicalFunctionScan            A
```

Comment:
Can we directly create the new operator (LogicalApplyAndMerge --- A) from the statement? I can see the current implementation requires few code modification. Does this transformation have any other benefits?

Reply:
I considered doing it. However, I feel that will require adding a lot more core. Via rules, it is easier to understand the changes, and we can fallback to lateral join whenever needed. What do you think?

Reply:
I can not think of any issues right now. We need to make sure that this rule applies first, because I assume all later rewrite rules (such as UDF reuse) will base on the new linear design.